File: src/udemy_tutorials_py/udemy_tutorials_py/add_two_ints_client_oop.py
#!/usr/bin/env python3
"""Example to communicate with the add_two_ints service without OOP."""
import sys
from functools import partial
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts


class AddTwoIntsClientNode(Node):
    """Client node with OOP."""
    def __init__(self):
        """Create an instance of the client node."""
        super().__init__("add_two_ints_client_oop")
        self.call_service(AddTwoInts.Request(a=3, b=4))

    def call_service(self, request):
        """Call the service."""
        client = self.create_client(AddTwoInts, "add_two_ints")
        self.get_logger().info("Client node has been created.")
        while not client.wait_for_service(timeout_sec=1.0):
            self.get_logger().warn("Service not available, waiting again...")
        self.get_logger().info("Service is available.")
        future = client.call_async(request)
        future.add_done_callback(
            partial(self.callback_call_service, request=request)
        )

    @property
    def logger(self):
        """Return the logger."""
        return self.get_logger()

    def callback_call_service(self, future, request):
        """Callback function for the service call."""
        response = future.result()
        self.logger.info(f"Request: {request.a} + {request.b}")
        self.logger.info(f"Result: {response.sum}")

def main(args=None):
    """Client node without OOP."""
    rclpy.init(args=args)
    # Create a node
    node = AddTwoIntsClientNode()
    rclpy.spin(node)
    # Destroy the node explicitly
    rclpy.shutdown()
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/udemy_tutorials_cpp/src/add_two_ints_client_oop.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"


/*
 * This example demostrates how to create a simple client node for the 
 * add_two_ints service using the OOP style.
 */

class AddTwoIntsClient : public rclcpp::Node
{
    public:
        AddTwoIntsClient() : Node("add_two_ints_client")
        {
            // Create a client for the add_two_ints service
            //
            // If we do the call directly, it will block the node and we will
            // not be able to do anything else. To avoid this, we can use a
            // thread to call the service.
            // call_add_two_ints_service(1, 2);
            //
            // Create a thread to call the service
            thread1_ = std::thread(std::bind(
                &AddTwoIntsClient::call_add_two_ints_service, this, 3, 4));
            thread2_ = std::thread(std::bind(
                &AddTwoIntsClient::call_add_two_ints_service, this, 5, 6));
        }
        void call_add_two_ints_service(int a, int b)
        {
            auto client = create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
            while(!client->wait_for_service(std::chrono::seconds(1))){
                RCLCPP_INFO(get_logger(), "Waiting for the server to be up...");
            }
            RCLCPP_INFO(get_logger(), "Server is up!!.");
            // Create a request message and set the values entered by the user
            auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
            request->a = a;
            request->b = b;
            // Call the service
            RCLCPP_INFO(get_logger(), "Sending service request:");
            RCLCPP_INFO(get_logger(), "  a: %ld", request->a);
            RCLCPP_INFO(get_logger(), "  b: %ld", request->b);
            auto future = client->async_send_request(request);
            // Wait for the future inside this future even if the node is spinning
            try {
                auto response = future.get();
                RCLCPP_INFO(get_logger(), "  Result = %ld", response->sum);
            } catch (const std::exception &e) {
                RCLCPP_ERROR(get_logger(), "Service call failed");
            }
        }
    private:
        std::thread thread1_;
        std::thread thread2_;
};

int main(int argc, char **argv)
{
    // Initialize the ROS 2 system
    rclcpp::init(argc, argv);
    // Create a node
    auto node = std::make_shared<AddTwoIntsClient>();
    // Spin the node and shutdown the ROS 2 system
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
